core/zoe/contracts/coveredCallSrcs.js

```diff
@@ -0,0 +1,71 @@
+import harden from '@agoric/harden';
+
+const areAssaysValid = array => array.length === 2;
+
+const isValidInitialOfferDesc = newOfferDesc => {
+  const makeHasOkRules = validRules => offer =>
+    validRules.every((rule, i) => rule === offer[i].rule, true);
+
+  const hasOkLength = array => array.length === 2;
+  const hasOkRules = offer =>
+    makeHasOkRules(['offerExactly', 'wantExactly'])(offer) ||
+    makeHasOkRules(['wantExactly', 'offerExactly'])(offer);
+  return hasOkLength(newOfferDesc) && hasOkRules(newOfferDesc);
+};
+
+const makeWantedOfferDescs = firstOfferDesc => {
+  const makeSecondOffer = firstOffer =>
+    harden([
+      {
+        rule: firstOffer[1].rule,
+        assetDesc: firstOffer[0].assetDesc,
+      },
+      {
+        rule: firstOffer[0].rule,
+        assetDesc: firstOffer[1].assetDesc,
+      },
+    ]);
+  return harden([makeSecondOffer(firstOfferDesc)]);
+};
+
+const isValidOfferDesc = (extentOps, offerDescToBeMade, offerDescMade) => {
+  const ruleEqual = (leftRule, rightRule) => leftRule.rule === rightRule.rule;
+
+  const extentEqual = (extentOp, leftRule, rightRule) =>
+    extentOp.equals(leftRule.assetDesc.extent, rightRule.assetDesc.extent);
+
+  const assayEqual = (leftRule, rightRule) =>
+    leftRule.assetDesc.label.assay === rightRule.assetDesc.label.assay;
+
+  // Check that two offers are equal in both their rules and their assetDescs
+  const offerEqual = (leftOffer, rightOffer) => {
+    const isLengthEqual = leftOffer.length === rightOffer.length;
+    if (!isLengthEqual) {
+      return false;
```

Comment:
With this line, I just noticed that the srcs APIs indicate failure with a `false` rather than something with diagnostic information. This can become frustrating when something that should have succeeded doesn't.

That's why the rejected promise carries a "reason". For synchronous cases where failure likely indicates that someone has a bug, I've generally used normal termination vs throwing an exception, such as with `insist`.